modules/FvwmIconMan/FvwmIconMan.man
.\" t
.\" @(#)FvwmIconMan.1	3/25/96
.TH FvwmIconMan 0.4 "Mar 25 1996"
.UC
.SH NAME
FvwmIconMan \- an Fvwm Icon Manager
.SH SYNOPSIS
FvwmIconMan is spawned by fvwm, so no command line invocation will work.

.SH DESCRIPTION
FvwmIconMan is an icon manager modeled after the TWM icon manager. 
The user may have multiple icon managers, each of which are armed with a list
of window types which it manages. For example, the user may have one manager
which lists only emacs windows, and another which lists everything else. You
may also specify what resolution each icon manager uses, for example, one
icon manager may manage windows on all desks, and another may manage only 
those on the current desk, or page.

Unlike the TWM icon manager, FvwmIconMan is limited in only having the icons 
listed vertically. Also unlike the TWM icon manager, it sports a three 
dimensional look (when the display depth is sufficient). 

You may bind fvwm functions to mouse buttondown events, and to the event 
where the mouse enters one of the buttons. For example, you could bind the
first mouse button to the Iconify function, and the second mouse button to
the Identify function.

FvwmIconMan can be set to display which window currently has the keyboard
focus, and by binding the select event (see below) to the fvwm Focus function,
you can emulate the TWM icon manager's behavior. Unfortunately, the Focus
function warps the cursor to the target window. This FvwmIconMan distribution
includes a patch to fvwm 2.0.41 which modifes the behavior of Focus so that
it takes an argmument to control whether or not it warps the cursor.

.SH INITIALIZATION
During initialization, FvwmIconMan searches though the fvwm configuration file
for the options which are described below. It is highly recommended that you
make FvwmIconMan be a sticky window. And if you want to make use of the 
followfocus option, and/or binding an action to Focus, then you should make
FvwmIconMan clicktofocus.

.SH INVOCATION
FvwmIconMan can be invoked by inserting the line 'Module FvwmIconMan' in the .fvwmrc file. If FvwmIconMan is to be spawned during fvwm's initialization,
then this line should be placed in the InitFunction and ResetFunction
declarations, or it can be bound to a menu, mouse button, or keystroke to
invoke it later. FvwmIconMan should be placed in the ModulePath (defined in
the .fvwmrc file) in order for fvwm to find it.

.SH CONFIGURATION OPTIONS
With the exception of the nummanagers option, all of the options may be
defined on a per-manager basis. So, for example, the user may have his emacs
manager with a red foreground, and his xterm manager with a blue one. A
configuration line may therefore have one of two forms:

.IP "*FvwmIconMan*optionname optionvalue"
To specify that the \fIoptionname\fP takes the value \fIoptionvalue\fP 
for all managers.
.IP "*FvwmIconMan*managerid*optionname optionvalue"
To specify that the option \fIoptionname\fP takes the value \fIoptionvalue\fP
for manager \fImanagerid\fP.

The following options may be specified:

.IP "*FvwmIconMan*nummanagers \fInum\fP"
\fInum\fP is a positive integer specifying the total number of icon managers.
Since FvwmIconMan would like to know how many managers there are before
handling any manager specific options, this should come first. Default is 1.

.IP "*FvwmIconMan*[id*]resolution \fIresolution\fP"
Specifies when the manager will display an entry for a certain
window. \fIresolution\fP may take one of the following values: global, desk,
or page. If global, then all windows of the appropriate type (see the show
and dontshow options below) will be shown. If desk, then only those windows 
on the current desk will be down. And if page, then only those windows on the 
current page will be shown. Default is global.

.IP "*FvwmIconMan*[id*]showtitle \fIboolean\fP"
If \fItrue\fP, then print the window title names in the
manager, if \fIfalse\fP, then print the icon names. Default is false.

.IP "*FvwmIconMan*[id*]font \fIfont\fP"
Specifies the font to be used for labeling the buttons.

.IP "*FvwmIconMan*[id*]geometry \fIgeometry\fP"
Specified the geometry of the manager. If the geometry is specified with a 
negative y coordinate, then the window manager will grow upwards. Otherwise,
it will shrink downwards.

.IP "*FvwmIconMan*[id*]foreground \fIforeground\fP"
Specifies the default foreground color.

.IP "*FvwmIconMan*[id*]background \fIbackground\fP"
Specifies the default background color.

.IP "*FvwmIconMan*[id*]plainbutton \fIstyle\fP [\fIforecolor\fP \fIbackcolor\fP]"
Specifies how normal buttons look. \fIstyle\fP may be one of \fIflat\fP,
\fIup\fP, or \fIdown\fP, and describes how the button is drawn. The
color options are both optional, and if not set, then the default
colors are used. If on a monochrome screen, then the \fIstyle\fP option is
ignored, but must still be set.

.IP "*FvwmIconMan*[id*]selectbutton \fIstyle\fP [\fIforecolor\fP \fIbackcolor\fP]"
Same as the plainbutton option, but specifies the look of buttons when the
mouse is over them.

.IP "*FvwmIconMan*[id*]focusbutton \fIstyle\fP [\fIforecolor\fP \fIbackcolor\fP]"
Same as the plainbutton option, but specifies the look of buttons whose
windows have the keyboard focus.

.IP "*FvwmIconMan*[id*]focusandselectbutton \fIstyle\fP [\fIforecolor\fP \fIbackcolor\fP]"
Same as the plainbutton option, but specifies the look of buttons which are
both selected, and have the keyboard focus.

.IP "*FvwmIconMan*[id*]action \fIevent\fP \fIresponse\fP[,\fIresponse...\fP]
Tells FvwmIconMan to send the commands in \fIresponse\fP to fvwm when 
\fIevent\fP is done. \fIEvent\fP may take on one of the values: select, 
click1, click2, or click3. Default is iconify for all three buttons and nop 
for select.  Select refers to the cursor entering a button in the manager.

The two following options control which windows get handled by which
managers. A manager can get two lists, one of windows to show, and one of
windows to ignore. If only the \fIshow\fP list is given, then that manager
will show only the windows in the list. If only the \fIdontshow\fP list is
given, then the manager will show all windows except those in the list. If
both lists are given, then a window will be shown if it is not in the
\fIdontshow\fP list, and in the \fIshow\fP list. And finally, if neither list
is given, then the manager will handle all windows. Each list is made up of
patterns of the form \fItype=pattern\fP, where type is one of \fIclass\fP,
\fIresource\fP, \fItitle\fP, or \fIicon\fP, and pattern is a regular
expression of the same format used in the fvwm style command. Quotes around
the pattern will be taken as part of the regular expression. If a window could
be handled by more than one manager, then the manager with the lowest id gets 
it.

.IP "*FvwmIconMan*[id*]show \fIpattern list\fP"
If a window matches one of the patterns in the list, then it may be handled
by this manager.

.IP "*FvwmIconMan*[id*]dontshow \fIpattern list\fP"
If a window matches one of the patterns in the list, then it may not be
handled by this manager.

.IP "*FvwmIconMan*[id*]followfocus \fIboolean\fP"
If \fIboolean\fP is \fItrue\fP, then the button appearance reflects
which window currently has focus.  Default is false.

.IP "*FvwmIconMan*[id*]sort \fIboolean\fP"
If \fIboolean\fP is \fItrue\fP, then the icon manager is kept sorted. Default
is true.

.SH SAMPLE CONFIGURATION
This first example is of a the simplest invocation of FvwmIconMan, which only
has one manager, and handles all windows:

.nf
.sp
##############################################################
# Load any modules which should be started during 
# fvwm initialization
ModulePath /usr/lib/X11/fvwm:/usr/bin/X11
Module FvwmIconMan

# Make FvwmIconMan title-bar-less, sticky, and give it an icon	
Style "Fvwm*"      Icon toolbox.xpm,NoTitle,NoHandles,Sticky
Style "FvwmIconMan" HandleWidth 5, Handles, BorderWidth 5


##############################################################
##############################################################
#Definitions used by the modules

*FvwmIconMan*nummanagers 1
*FvwmIconMan*resolution  global
*FvwmIconMan*background  slategrey
*FvwmIconMan*foreground  white
*FvwmIconMan*font        7x13
*FvwmIconMan*geometry    194x100-0+73
.sp
.fi

This example is my personal configuration. It has two managers, one for emacs
and one for everything else, minus things with no icon title. Only windows on
the current page are displayed. Notice the argument to Focus in the select
action. This only works if you apply the included patch for fvwm. A nonzero
argument to focus, or no argument results in Focus warping the cursor to the
target window, and the value of 0 makes Focus not warp the cursor. Note how
the geometry and show options are specified per manager, and the others are
common to all:

.nf
.sp
*FvwmIconMan*numManagers 2
*FvwmIconMan*Resolution  page
*FvwmIconMan*background  steelblue
*FvwmIconMan*foreground  white
*FvwmIconMan*font        7x13
*FvwmIconMan*action      click1 Iconify
*FvwmIconMan*action      click2 Iconify
*FvwmIconMan*action      click3 Module "FvwmIdent" FvwmIdent
*FvwmIconMan*action      select Focus 0
*FvwmIconMan*showtitle   false
*FvwmIconMan*followfocus true
*FvwmIconMan*sort        true
*FvwmIconMan*plainbutton          up white steelblue
*FvwmIconMan*selectbutton         down white steelblue
*FvwmIconMan*focusbutton          up white brown
*FvwmIconMan*focusandselectButton down white brown

*FvwmIconMan*1*geometry   194x100-204-90
*FvwmIconMan*1*show       resource=emacs resource=gemacs

*FvwmIconMan*2*geometry   194x100-0-90
*FvwmIconMan*2*dontshow   icon=Untitled

.sp
.fi

.SH UNFINISHED BUSINESS
There is one bug that I know of. A honest to goodness solution to this would
be appreciated. When an icon manager is set to grow upwards, on some machines
it may wander occasionally.

When a manager is in page resolution, and the page is switched, the manager
updates the windows after every fvwm event comes in, causing it to rapidly
grow and shrink until it finally settles into the new configuration. This
doesn't happen when in desk resolution.

It doesn't handle windows without resource names as gracefully as it should.

.SH AUTHOR
Brady Montz (bradym@cs.arizona.edu).

.SH THANKS
Thanks to David Berson (berson@cs.pitt.edu).


Changelog
Wed Feb 4 1998 Emmanuel Rouat <rouat@dsys.ceng.cea.fr>
	* created system.fvwm95rc.in
	
Wed Jan 28 1998 Emmanuel Rouat <rouat@dsys.ceng.cea.fr>
	* Makefile of FvwmConsole corrected (FvwmConsoleC was missing)

Tue Jan 27 1998 Emmanuel Rouat <rouat@dsys.ceng.cea.fr>
	* Makefiles updated - should compile and install smoothly now

Thu Jan 22 1998 Emmanuel Rouat <rouat@dsys.ceng.cea.fr>
	* Scripts for FvwmScripts adapted
	
Tue Jan 20 1998 Emmanuel Rouat <rouat@dsys.ceng.cea.fr>
	* sample/system.fvwm95rc lots of tweaking (on going work)
	* FvwmScript module added (from fvwm-2.0.46)
	* FvwmForm module updated ( ''    ''    '' )
	* FvwmWharf module added  ( ''    ''    '' )

Tue Jan 6 1998 Emmanuel Rouat <rouat@dsys.ceng.cea.fr>
	* modules/FvwmTaskBar 'Enhanced FvwmTasbar' instead of regular one

modules/FvwmIconMan/FvwmIconMan.c
#include "FvwmIconMan.h"
#include "../../fvwm/module.h"
#include <fvwm/fvwmlib.h>

static int fd_width;
static FILE *console = NULL;
static char *VERSION = "0.4";

#if !defined (__GNUC__) || defined (PRINT_DEBUG)
char *actionNames[] = {
  "click1",
  "click2",
  "click3",
  "select"
};
#endif

typedef enum {
  READ_LINE = 1,
  READ_OPTION = 2,
  READ_ARG = 4,
  READ_REST_OF_LINE = 12
} ReadOption;

void Free (void *p)
{
  if (p != NULL)
    free (p);
}

static int iswhite (char c)
{
  if (c == ' ' || c == '\t' || c == '\0')
    return 1;
  return 0;
}

static void skip_space (char **p)
{
  while (**p == ' ' || **p == '\t')
    (*p)++;
}

#if FVWM_VERSION == 1
static FILE *config_fp = NULL;
#endif

static int init_config_file (char *file)
{
#if FVWM_VERSION == 1
  if ((config_fp = fopen (file, "r")) == NULL)  {
    ConsoleMessage ("Couldn't open file: %s\n", file);
    return 0;
  }
#endif
  return 1;
}

static void close_config_file (void)
{
#if FVWM_VERSION == 1
  if (config_fp)
    fclose (config_fp);
#endif
}

static int GetConfigLineWrapper (int *fd, char **tline)
{
#if FVWM_VERSION == 1

  static char buffer[1024];
  char *temp;

  if (fgets (buffer, 1024, config_fp)) {
    *tline = buffer;
    temp = strchr (*tline, '\n');
    if (temp) {
      *temp = '\0';
    }
    else {
      fprintf (stderr, "line too long\n");
      exit (1);
    }
    return 1;
  }

#else

  char *temp;

  GetConfigLine (fd, tline);
  if (*tline) {
    temp = strchr (*tline, '\n');
    if (temp) {
      *temp = '\0';
    }
    return 1;
  }

#endif

  return 0;
}

static char *read_next_cmd (ReadOption flag)
{
  static ReadOption status;
  static char *buffer;
  static char *retstring, displaced, *cur_pos;

  retstring = NULL;
  if (flag != READ_LINE && !(flag & status))
    return NULL;

  switch (flag) {
  case READ_LINE:
    while (GetConfigLineWrapper (Fvwm_fd, &buffer)) {
      cur_pos = buffer;
      skip_space (&cur_pos);
      if (!strncasecmp (Module, cur_pos, ModuleLen)) {
        retstring = cur_pos;
        cur_pos += ModuleLen;
        displaced = *cur_pos;
        if (displaced == '*') 
          status = READ_OPTION;
        else if (displaced == '\0')
          status = READ_LINE;
        else if (iswhite (displaced))
          status = READ_ARG;
        else 
          status = READ_LINE;
        break;
      }
    }
    break;

  case READ_OPTION:
    *cur_pos = displaced;
    retstring = ++cur_pos;
    while (*cur_pos != '*' && !iswhite (*cur_pos)) 
      cur_pos++;
    displaced = *cur_pos;
    *cur_pos = '\0';
    if (displaced == '*')
      status = READ_OPTION;
    else if (displaced == '\0')
      status = READ_LINE;
    else if (iswhite (displaced))
      status = READ_ARG;
    else 
      status = READ_LINE;
    break;

  case READ_ARG:
    *cur_pos = displaced;
    skip_space (&cur_pos);
    retstring = cur_pos;
    while (!iswhite (*cur_pos))
      cur_pos++;
    displaced = *cur_pos;
    *cur_pos = '\0';
    if (displaced == '\0')
      status = READ_LINE;
    else if (iswhite (displaced))
      status = READ_ARG;
    else 
      status = READ_LINE;
    break;

  case READ_REST_OF_LINE:
    status = READ_LINE;
    *cur_pos = displaced;
    skip_space (&cur_pos);
    retstring = cur_pos;
    break;
  }

  if (retstring && retstring[0] == '\0')
    retstring = NULL;

  return retstring;
}

int OpenConsole()
{
#ifdef PRINT_CONSOLE
  if ((console=fopen("/dev/console","w"))==NULL) {
    fprintf(stderr,"%s: cannot open console\n",Module);
    return 0;
  }
#endif
  return 1;
}

void ShutMeDown (int flag)
{
  ConsoleDebug ("Bye Bye\n");
  exit (flag);
}

void ConsoleMessage(char *fmt, ...)
{
#ifdef PRINT_CONSOLE
  va_list args;
  FILE *filep;
  if (console==NULL) filep=stderr;
  else filep=console;
  va_start(args,fmt);
  vfprintf(filep,fmt,args);
  va_end(args);
#endif
}

#if defined(PRINT_DEBUG) || !defined(__GNUC__)
void ConsoleDebug (char *fmt, ...) {
#ifdef PRINT_CONSOLE
#ifdef PRINT_DEBUG
  va_list args;
  FILE *filep;
  if (console==NULL) filep=stderr;
  else filep=console;
  va_start(args,fmt);
  vfprintf(filep,fmt,args);
  va_end(args);
#endif
#endif
}
#endif

void DeadPipe (int nothing)
{
  ConsoleDebug ("Bye Bye\n");
  ShutMeDown (0);
}

void SendFvwmPipe (char *message,unsigned long window)
{
  char *hold,*temp,*temp_msg;
  hold=message;

  while(1) {
    temp=strchr(hold,',');
    if (temp!=NULL) {
      temp_msg= (char *)safemalloc(temp-hold+1);
      strncpy(temp_msg,hold,(temp-hold));
      temp_msg[(temp-hold)]='\0';
      hold=temp+1;
    } else temp_msg=hold;

    SendText(Fvwm_fd, temp_msg, window);

    if(temp_msg!=hold) free(temp_msg);
    else break;
  }
}

static void main_loop (void)
{
  fd_set readset;
  struct timeval tv;

  while(1) {
    FD_ZERO( &readset);
    FD_SET (Fvwm_fd[1],&readset);
    FD_SET (x_fd,&readset);
    tv.tv_sec=0;
    tv.tv_usec=0;
    if (!select(fd_width,&readset,NULL,NULL,&tv)) {
      FD_ZERO (&readset);
      FD_SET (Fvwm_fd[1],&readset);
      FD_SET (x_fd, &readset);
      select(fd_width,&readset,NULL,NULL,NULL);
    }

    if (FD_ISSET (x_fd, &readset) || XPending (theDisplay)) {
      xevent_loop();
    }
    if (FD_ISSET(Fvwm_fd[1],&readset)) {
      ReadFvwmPipe();
    }
  }
}

static int extract_int (char *p, int *n)
{
  char *s;

  for (s = p; *s; s++) {
    if (*s < '0' || *s > '9') {
      return 0;
    }
  }

  *n = atoi (p);

  return 1;
}      

char *copy_string (char **target, char *src)
{
  int len = strlen (src);

  if (*target)
    Free (*target);
  
  *target = (char *)safemalloc ((len + 1) * sizeof (char));
  strcpy (*target, src);
  return *target;
}

static char *conditional_copy_string (char **s1, char *s2)
{
  if (*s1)
    return *s1;
  else
    return copy_string (s1, s2);
}

#define SET_MANAGER(manager,field,value)                           \
   do {                                                            \
     int id = manager;                                             \
     if (id == -1) {                                               \
       for (id = 0; id < globals.num_managers; id++) {             \
	 globals.managers[id]. field = value;                     \
       }                                                           \
     }                                                             \
     else if (id < globals.num_managers) {                         \
       globals.managers[id]. field = value;                       \
     }                                                             \
     else {                                                        \
       ConsoleMessage ("Internal error in SET_MANAGER: %d\n", id); \
     }                                                             \
   } while (0)

static void read_in_resources (char *file)
{
  char *p, *q;
  int i, n, manager;
  char *option1;
  Resolution r = SHOW_GLOBAL;
  ButtonState state;

  if (!init_config_file (file))
    return;

  while ((p = read_next_cmd (READ_LINE))) {
    ConsoleDebug ("line: %s\n", p);
    option1 = read_next_cmd (READ_OPTION);
    if (option1 == NULL)
      continue;

    ConsoleDebug ("option1: %s\n", option1);
    if (!strcasecmp (option1, "nummanagers")) {
      p = read_next_cmd (READ_ARG);
      if (!option1) {
	ConsoleMessage ("Error in input file\n");
	continue;
      }
      if (extract_int (p, &n) == 0) {
	ConsoleMessage ("This is not a number: %s\n", p);
	continue;
      }
      if (n > 0) {
	allocate_managers (n);
	ConsoleDebug ("num managers: %d\n", n);
      }
      else {
	ConsoleMessage ("You can't have zero managers. I'll give you one.\n");
	allocate_managers (1);
      }
    }
    else {
      /* these all can specify a specific manager */

      if (globals.managers == NULL) {
	ConsoleDebug ("I'm assuming you only want one manager\n");
	allocate_managers (1);
      }

      manager = 0;

      if (option1[0] >= '0' && option1[0] <= '9') {
	if (extract_int (option1, &manager) == 0 || 
	    manager <= 0 || manager > globals.num_managers) {
	  ConsoleMessage ("This is not a valid manager: %s.\n", option1);
	  manager = 0;
	}
	option1 = read_next_cmd (READ_OPTION);
	if (!option1) {
	  ConsoleMessage ("Error in input file\n");
	  continue;
	}
      }

      manager--; /* -1 means global */

      ConsoleDebug ("Applying %s to manager %d\n", option1, manager);

      if (!strcasecmp (option1, "resolution")) {
	p = read_next_cmd (READ_ARG);
	if (!p) {
	  ConsoleMessage ("Error in input file\n");
	  continue;
	}
	ConsoleDebug ("resolution: %s\n", p);
	if (!strcasecmp (p, "global"))
	  r = SHOW_GLOBAL;
	else if (!strcasecmp (p, "desk"))
	  r = SHOW_DESKTOP;
	else if (!strcasecmp (p, "page"))
	  r = SHOW_PAGE;

	SET_MANAGER (manager, res, r);
      }
      else if (!strcasecmp (option1, "font")) {
	p = read_next_cmd (READ_ARG);
	if (!p) {
	  ConsoleMessage ("Error in input file\n");
	  continue;
	}
	ConsoleDebug ("font: %s\n", p);

	SET_MANAGER (manager, fontname, 
		     copy_string (&globals.managers[id].fontname, p));
      }
      else if (!strcasecmp (option1, "geometry")) {
	p = read_next_cmd (READ_ARG);
	if (!p) {
	  ConsoleMessage ("Error in input file\n");
	  continue;
	}

	SET_MANAGER (manager, geometry, 
		     copy_string (&globals.managers[id].geometry, p));
      }
      else if (!strcasecmp (option1, "dontshow")) {
	p = read_next_cmd (READ_ARG);
	if (!p) {
	  ConsoleMessage ("Error in input_file\n");
	  continue;
	}
	do {
	  ConsoleDebug ("dont show: %s\n", p);
	  if (manager == -1) {
	    int i;
	    for (i = 0; i < globals.num_managers; i++)
	      add_to_stringlist (&globals.managers[i].dontshow, p);
	  }
	  else {
	    add_to_stringlist (&globals.managers[manager].dontshow, p);
	  }
	  p = read_next_cmd (READ_ARG);
	} while (p);
      }
      else if (!strcasecmp (option1, "show")) {
	p = read_next_cmd (READ_ARG);
	if (!p) {
	  ConsoleMessage ("Error in input_file\n");
	  continue;
	}
	do {
	  ConsoleDebug ("show: %s\n", p);
	  if (manager == -1) {
	    int i;
	    for (i = 0; i < globals.num_managers; i++)
	      add_to_stringlist (&globals.managers[i].show, p);
	  }
	  else {
	    add_to_stringlist (&globals.managers[manager].show, p);
	  }
	  p = read_next_cmd (READ_ARG);
	} while (p);
      }
      else if (!strcasecmp (option1, "background")) {
	p = read_next_cmd (READ_ARG);
	if (!p) {
	  ConsoleMessage ("Error in input file\n");
	  continue;
	}
	ConsoleDebug ("default background: %s\n", p);

        for ( i = 0; i < NUM_CONTEXTS; i++ )
	  SET_MANAGER (manager, backColorName[i], 
	    conditional_copy_string (&globals.managers[id].backColorName[i], 
				     p));
      }
      else if (!strcasecmp (option1, "foreground")) {
	p = read_next_cmd (READ_ARG);
	if (!p) {
	  ConsoleMessage ("Error in input file\n");
	  continue;
	}
	ConsoleDebug ("default foreground: %s\n", p);

        for ( i = 0; i < NUM_CONTEXTS; i++ )
	SET_MANAGER (manager, foreColorName[i], 
           conditional_copy_string (&globals.managers[id].foreColorName[i], 
				    p));
      }
      else if (!strcasecmp (option1, "action")) {
	p = read_next_cmd (READ_ARG);
	if (!p) {
	  ConsoleMessage ("Error in input file: need arguments to action\n");
	  continue;
	}
	if (!strcasecmp (p, "click1")) {
	  i = CLICK1;
	}
	else if (!strcasecmp (p, "click2")) {
	  i = CLICK2;
	}
	else if (!strcasecmp (p, "click3")) {
	  i = CLICK3;
	}
        else if (!strcasecmp (p, "select")) {
           i = SELECT;
        }
	else {
	  ConsoleMessage ("Error in input file: this isn't a valid action name: %s\n", p);
	  continue;
	}

	q = read_next_cmd (READ_REST_OF_LINE);
	if (!q) {
	  ConsoleMessage ("Error in input file: need an action\n");
	  continue;
	}
	ConsoleDebug ("Setting action for %s: %s\n", actionNames[i], q);
	SET_MANAGER (manager, actions[i], 
		     copy_string (&globals.managers[id].actions[i], q));
      }
      else if (!strcasecmp (option1, "showtitle")) {
	p = read_next_cmd (READ_ARG);
	if (!p) {
	  ConsoleMessage ("Error in input file: need argument to showtitle\n");
	  continue;
	}
	if (!strcasecmp (p, "true")) {
	  i = 1;
	}
	else if (!strcasecmp (p, "false")) {
	  i = 0;
	}
	else {
	  ConsoleMessage ("Error in input file. What is this: %s?\n", p);
	  continue;
	}
	ConsoleDebug ("Setting showtitle to: %d\n", i);
	SET_MANAGER (manager, use_titlename, i);
      }
      else if (!strcasecmp (option1, "plainButton")  ||
               !strcasecmp (option1, "selectButton")  ||
               !strcasecmp (option1, "focusButton")   ||
	       !strcasecmp (option1, "focusandselectButton")) {
         if ( !strcasecmp (option1, "plainButton") )
            i = PLAIN_CONTEXT;
         else if ( !strcasecmp (option1, "selectButton") )
            i = SELECT_CONTEXT;
         else if (!strcasecmp (option1, "focusandselectButton"))
	   i = FOCUS_SELECT_CONTEXT;
	 else
            i = FOCUS_CONTEXT;
         p = read_next_cmd (READ_ARG);
         if (!p) {
            ConsoleMessage ("Error in input file: need argument to %s\n",
			    option1);
            continue;
         }
         else if (!strcasecmp (p, "flat")) {
            state = BUTTON_FLAT;
         }
         else if (!strcasecmp (p, "up")) {
            state = BUTTON_UP;
         }
         else if (!strcasecmp (p, "down")) {
            state = BUTTON_DOWN;
         }
         else {
           ConsoleMessage ("Error in input file: this isn't a valid button state: %s\n", p);
           continue;
         }
	 ConsoleDebug ("Setting buttonState[%s] to %s\n", 
		       contextDefaults[i].name, p);
         SET_MANAGER (manager, buttonState[i], state);

            /* check for optional fore color */
         p = read_next_cmd (READ_ARG);
         if ( !p )
            continue;
         
         ConsoleDebug ("Setting foreColorName[%s] to %s\n", 
		       contextDefaults[i].name, p);
	 SET_MANAGER (manager, foreColorName[i], 
		      copy_string (&globals.managers[id].foreColorName[i], p));

            /* check for optional back color */
         p = read_next_cmd (READ_ARG);
         if ( !p )
            continue;
         
         ConsoleDebug ("Setting backColorName[%s] to %s\n",
		       contextDefaults[i].name, p);
	 SET_MANAGER (manager, backColorName[i], 
		      copy_string (&globals.managers[id].backColorName[i], p));
      }
      else if (!strcasecmp (option1, "followfocus")) {
	p = read_next_cmd (READ_ARG);
	if (!p) {
	  ConsoleMessage ("Error in input file: need argument to followfocus\n");
	  continue;
	}
	if (!strcasecmp (p, "true")) {
	  i = 1;
	}
	else if (!strcasecmp (p, "false")) {
	  i = 0;
	}
	else {
	  ConsoleMessage ("Error in input file. What is this: %s?\n", p);
	  continue;
	}
	ConsoleDebug ("Setting followfocus to: %d\n", i);
	SET_MANAGER (manager, followFocus, i);
      }
      else if (!strcasecmp (option1, "sort")) {
	p = read_next_cmd (READ_ARG);
	if (!p) {
	  ConsoleMessage ("Error in input file: need argument to sort\n");
	  continue;
	}
	if (!strcasecmp (p, "true")) {
	  i = 1;
	}
	else if (!strcasecmp (p, "false")) {
	  i = 0;
	}
	else {
	  ConsoleMessage ("Error in input file. What is this: %s?\n", p);
	  continue;
	}
	ConsoleDebug ("Setting sort to: %d\n", i);
	SET_MANAGER (manager, sort, i);
      }
      else {
	ConsoleMessage ("Unknown option: %s\n", p);
      }
    }
  }

  if (globals.managers == NULL) {
    ConsoleDebug ("I'm assuming you only want one manager\n");
    allocate_managers (1);
  }
  print_managers();
  close_config_file();
}

int main (int argc, char **argv)
{
  char *temp, *s;
  FvwmPacketBody temppacket;

#ifdef I18N
  setlocale(LC_CTYPE, "");
#endif

  if ((char *)&temppacket.add_config_data.window_gravity - 
      (char *)&temppacket != 21 * sizeof (long)) {
    fprintf (stderr, "Bad packet structure. Can't run. Sorry\n");
    exit (1);
  }

  OpenConsole();
  init_globals();
  init_winlists();
  
  temp = argv[0];
  s = strrchr (argv[0], '/');
  if (s != NULL)
    temp = s + 1;

  if((argc != 6) && (argc != 7)) {
    fprintf(stderr,"%s Version %s should only be executed by fvwm!\n",Module,
      VERSION);
    ShutMeDown (1);
  }

  Fvwm_fd[0] = atoi(argv[1]);
  Fvwm_fd[1] = atoi(argv[2]);
  init_display();
  init_boxes();
  
  signal (SIGPIPE, DeadPipe);  

  read_in_resources (argv[3]);
  
  assert (globals.managers);
  fd_width = GetFdWidth();

  SetMessageMask(Fvwm_fd,M_CONFIGURE_WINDOW | M_RES_CLASS | M_RES_NAME |
                 M_ADD_WINDOW | M_DESTROY_WINDOW | M_ICON_NAME |
                 M_DEICONIFY | M_ICONIFY | M_END_WINDOWLIST |
                 M_NEW_DESK | M_NEW_PAGE | M_FOCUS_CHANGE | M_WINDOW_NAME);

  SendInfo (Fvwm_fd, "Send_WindowList", 0);
  
  main_loop();

  ConsoleMessage ("Shouldn't be here\n");
  
  return 0;
}
